text/0000-offline-node-improved-trajectories-bags.md

```diff
@@ -0,0 +1,90 @@
+# Improved handling of multiple trajectories and bags in the offline node
+
+## Summary
+[summary]: #summary
+
+The offline node currently supports multiple sequential (non-concurrent) trajectories via loading multiple bags, where each bag is a separate trajectory.
+This RFC aims to give the user greater flexibility when dealing with multiple robots, multiple (possibly concurrent) trajectories and multiple bags.
+
+## Motivation
+[motivation]: #motivation
+
+Let's give an overview of all possibilities how a dataset layout and content can look like:
+
+| Robot plurality                                                         | Trajectory plurality                                 | Bag plurality                                                          |
+|-------------------------------------------------------------------------|------------------------------------------------------|------------------------------------------------------------------------|
+| __A.__ Single robot                                                     | __1.__ Single trajectory                             | __a.__ Single bag                                                      |
+| __B.__ Multiple identical robots  (using the same sensor topics/frames) | __2.__ Multiple trajectories, necessarily sequential | __b.__ Multiple continuous bags  (like a single bag) - a bag aggregate |
+| __C.__ Multiple different robots  (different sensor topics/frames)      | __3.__ Multiple trajectories,  possibly concurrent   | __c.__ Multiple bags                                                   |
+
+The online node currently supports the following combinations:
+__A1__, __A/B 2__ and __C 2/3__ (using `StartTrajectory` and `FinalizeTrajectory` service calls)
+
+The offline node currently supports: __A1a__, __A/B 2c__.
+
+The online node can never support __B3__, because only single robot's data can be published on a topic at a time.
+Also, __B3__ and __a/b__ is impossible for the same reasons.
+
+Note that some combinations are invalid, such as __A3__ (a single robot can't be on multiple concurrent trajectories), and __B/C 1__ (there can't be a single trajectory if there are multiple robots).
+Furthermore, __A/B 2 a/b__ would require us to be able to specify when a trajectory stops and begins within the same bag aggregate and on the same range data topics, which would be too complicated and out of scope.
+
+The aim of the RFC is to enable the user to use the offline node for __C2/3__ like in the online node, to preserve the ability for __A/B 2c__, and to possibly offer additional functionality: for example, __b__, and __B3c__, which is otherwise impossible with the online node.
```

Comment:
What I'm missing here is a true motivation, i.e. why are you interested in seeing trajectories being built at the same time in the offline node. Especially given that it has an unclear effect on the end result (all data will be processed whether in parallel or sequential) and it does not change the computational effort. For the main use case of the offline node right now, i.e. process data as quickly as possible and output the result, this seems irrelevant.

Reply:
You are right about this part - being able to watch concurrent trajectories being built at the same time does not influence the end result in any way besides potentially increasing the coolness factor. However, it's possible to do it with the online node. Why not have the ability to do the same thing in the offline node, at increased processing speed? I believe that the offline node should offer everythihg that the online node can do, but at greater speed.

However, if I am correct, the offline node does not currently support concurrent trajectories _at all_. Or different robots.

The true motivation is stated at the beginning - to make the offline node more flexible and possible to be used in these cases. Someone who has bagged a swarm of robots and wants to try out cooperative SLAM with Cartographer can also be interested in crunching the bags as fast as possible.

Reply:
I guess what you explained now is what I was missing. Add these use cases (concurrent trajectories being built for additional coolness, different robots) at the very beginning. This seems to be the motivation to me. The other discussion is already the way how to get there.